File: esw-ocs/esw-ocs-dsl-kt/src/main/kotlin/esw/ocs/dsl/lowlevel/CswServices.kt
/*
 * Copyright (c) 2022 Thirty Meter Telescope International Observatory
 * SPDX-License-Identifier: Apache-2.0
 */

package esw.ocs.dsl.lowlevel

import akka.actor.typed.ActorSystem
import akka.actor.typed.SpawnProtocol
import csw.alarm.api.javadsl.IAlarmService
import csw.config.api.javadsl.IConfigClientService
import csw.config.client.javadsl.JConfigClientFactory
import csw.database.DatabaseServiceFactory
import csw.event.api.javadsl.IEventPublisher
import csw.event.api.javadsl.IEventService
import csw.event.api.javadsl.IEventSubscriber
import csw.location.api.javadsl.ILocationService
import csw.location.client.javadsl.JHttpLocationServiceFactory
import csw.logging.api.javadsl.ILogger
import csw.time.scheduler.TimeServiceSchedulerFactory
import csw.time.scheduler.api.TimeServiceScheduler
import esw.ocs.dsl.script.StrandEc
import esw.ocs.impl.script.ScriptContext

/**
 * An Interface which represents Collection of csw-services (Location, EventService etc.)
 *
 */
interface CswServices {

    companion object {
        /**
         * Creates an instance of CswServices
         *
         * @param ctx - an instance of ScriptContext
         * @param strandEc - an instance of strandEc (mostly single threaded execution context)
         * @return an instance of [CswServices]
         */
        internal fun create(ctx: ScriptContext, strandEc: StrandEc): CswServices = object : CswServices {
            private val actorSystem: ActorSystem<SpawnProtocol.Command> = ctx.actorSystem()

            override val logger: ILogger = ctx.jLogger()
            override val alarmService: IAlarmService = ctx.alarmService()
            override val eventService: IEventService = ctx.eventService()
            override val eventPublisher: IEventPublisher by lazy { eventService.defaultPublisher() }
            override val eventSubscriber: IEventSubscriber by lazy { eventService.defaultSubscriber() }
            override val locationService: ILocationService by lazy { JHttpLocationServiceFactory.makeLocalClient(actorSystem) }
            override val configService: IConfigClientService by lazy { JConfigClientFactory.clientApi(actorSystem, locationService) }
            override val timeService: TimeServiceScheduler by lazy { TimeServiceSchedulerFactory(actorSystem.scheduler()).make(strandEc.ec()) }
            override val databaseServiceFactory: DatabaseServiceFactory by lazy { DatabaseServiceFactory(actorSystem) }
        }
    }

    val logger: ILogger
    val alarmService: IAlarmService
    val eventService: IEventService
    val eventPublisher: IEventPublisher
    val eventSubscriber: IEventSubscriber
    val locationService: ILocationService
    val configService: IConfigClientService
    val timeService: TimeServiceScheduler
    val databaseServiceFactory: DatabaseServiceFactory
}

File: esw-ocs/esw-ocs-dsl-scala/src/main/scala/esw/ocs/dsl2/lowlevel/CswServices.scala
package esw.ocs.dsl2.lowlevel

import akka.actor.typed.{ActorSystem, SpawnProtocol}
import csw.alarm.api.scaladsl.AlarmService
import csw.config.api.scaladsl.ConfigClientService
import csw.config.client.scaladsl.ConfigClientFactory
import csw.database.DatabaseServiceFactory
import csw.event.api.scaladsl.{EventPublisher, EventService, EventSubscriber}
import csw.location.api.scaladsl.LocationService
import csw.location.client.scaladsl.HttpLocationServiceFactory
import csw.logging.api.scaladsl.Logger
import esw.ocs.dsl.script.StrandEc
import esw.ocs.impl.script.ScriptContext
import csw.time.scheduler.TimeServiceSchedulerFactory
import csw.time.scheduler.api.TimeServiceScheduler

class CswServices(ctx: ScriptContext, strandEc: StrandEc) {
  val alarmService: AlarmService         = ctx.alarmService.asScala
  val eventService: EventService         = ctx.eventService.asScala
  val eventPublisher: EventPublisher     = eventService.defaultPublisher
  val eventSubscriber: EventSubscriber   = eventService.defaultSubscriber
  val locationService: LocationService   = HttpLocationServiceFactory.makeLocalClient(ctx.actorSystem)
  val configService: ConfigClientService = ConfigClientFactory.clientApi(ctx.actorSystem, locationService)
  val timeService: TimeServiceScheduler  = TimeServiceSchedulerFactory()(ctx.actorSystem.scheduler).make()(strandEc.ec)
  val databaseServiceFactory: DatabaseServiceFactory = new DatabaseServiceFactory(ctx.actorSystem)
}
